Rewrite to version 5.0.0

piotr-oles opened the pull request:
As I promised a few weeks ago, I've performed a rewrite of the plugin to pay off technical debt and open this plugin for future development. This is a draft PR to show you the progress.

## Tasks:
- [x] Pay off technical debt
- [x] Support for TypeScript Watch API controlled by webpack (#362, #332, #330, #236, #211)
- [x] Support for TypeScript Solution Builder API (projectReferences) (#392, #328, #187)
- [x] Support for TypeScript Incremental API (partially implemented by using SolutionBuilder) (#238)
- [x] Support for Yarn PnP
- [x] Support for EsLint
- [x] TypeScript extensions infrastructure
- [x] Support for Vue.js (TypeScript Watch API) (#219)
- [x] Improve Vue template compiler error message (#209)
- [x] Documentation (#298, #36 - add information about `importsNotUsedAsValues: preserve`)
- [ ] **Unit Tests**
- [x] **E2E Tests**
- [x] Reevaluate supported versions (TypeScript >=2.7.0, Webpack >=4.0.0, Node.js >=10)  (#257)
- [x] Properly closing services on out of memory error (#387)
- [x] Inform users on out of memory error that its probably bug in the TypeScript itself
- [x] Support for Webpack 5 (#383)
- [ ] Support for Webpack multi-compiler (#229)
- [x] Ensure that types are properly exposed (#359, #339, #188)
- [x] Using the external format for Compiler Options (#393, #112)
- [x] Unified issues processing (include, exclude, issues hook) (#396, #374, #340, #232, #216)
- [x] External configuration (#397)
- [ ] Review breaking changes and help with migration for major packages that cooperates with this plugin (ts-loader, create-react-app, pnp-webpack-plugin - please let me know if I should consider other packages too)
- [x] Integrate with webpack-dev-server with async: true mode (#180)
- [x] Set default skipLibCheck: true option (#128)
- [x] Don't pollute --json output of stats (#101)
- [x] Better debugging experience (for example: `debug: true` flag) (#59)

This release should fix also: #309, #303, #273

## Plan:
 1. release first alpha version when unit tests will be ready to have at least minimal test coverage.
 2. release second alpha version when integration test will be ready to have it stable
 3. update documentation
 3. go through the rest of checklist items, implement it in a separate PRs and release

I'm not sure if I will go through all these points before releasing stable 5.0.0 - some of these points can be implemented later 🚀

piotr-oles committed:
refactor: make CancellationToken independent from TypeScript

BREAKING CHANGE: 🧨 Removed `throwIfCancellationRequested` method from the CancellationToken
passed in the `cancel` hook

piotr-oles committed:
feat: initial implementation of the new architecture

piotr-oles committed:
fix: use 'afterCompile' hook to fix regression described in #57

johnnyreilly commented on src/error/OperationCancelledError.ts:
```diff
@@ -0,0 +1,11 @@
+  constructor(message: string) {
+    super(message);
+
+    Object.setPrototypeOf(this, OperationCancelledError.prototype);
```
What benefits does `Object.setPrototypeOf` provide? I'm someone who tends to avoid inheritance in general and so I'm not so clear on this I'm afraid. Would love to understand!

piotr-oles replied:
It's only because I'm extending base `Error` class - normally I would also avoid inheritence. The explanation of why it's needed is here: https://stackoverflow.com/a/41102306/11506703

johnnyreilly replied:
Interesting! I wonder if we need to transpile to es5 anymore given the minimum supported node version is 8.6? Doesn't that mean es6 is supported anyway?

piotr-oles replied:
Currently, we support node >=6.11.5. We can upgrade to node 8 as it will be major release and node 6 is no longer supported :)

johnnyreilly replied:
Cool - does that mean we can change our emit to es6 and remove `Object.setPrototypeOf(this, OperationCancelledError.prototype)`? As I understand it it's only required for es5 transpilation reasons?

By the way, as of December node 8 is also no longer supported too! ☺️

https://blog.risingstack.com/update-nodejs-8-end-of-life-no-support/

piotr-oles replied:
Yes, it seems that we don't need that for node 8 :) I will change the minimum node version

johnnyreilly commented on src/ForkTsCheckerWebpackPluginOptions.json:
```diff
@@ -0,0 +1,265 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
```
Did you write this JSON schema file by hand or is it generated somehow? What's the advantage of this file? Does it help power tooling? I've never used JSON schema myself and I'm intrigued!

piotr-oles replied:
I wrote it by hand :) I wanted to follow the webpack's guideline: https://webpack.js.org/contribute/writing-a-plugin/#basic-plugin-architecture
The purpose of this file is to validate configuration and give feedback if something is configured not properly.

johnnyreilly replied:
Ah cool!

johnnyreilly commented on src/hooks/tapDoneToAsyncGetIssues.ts:
```diff
@@ -0,0 +1,70 @@
+      issues = await report;
+    } catch (error) {
+      if (!(error instanceof OperationCancelledError)) {
+        hooks.error.call(error, stats.compilation);
```
Minor thing, I wonder if this might be easier to read if the cases were reversed:

```suggestion
	      if (error instanceof OperationCancelledError) {
```

For my money that's easier to understand than `(!(error...`

Just a thought

piotr-oles replied:
:+1:

johnnyreilly commented on src/hooks/tapAfterCompileToGetIssues.ts:
```diff
@@ -0,0 +1,69 @@
+
+        issues = await state.report;
+      } catch (error) {
+        if (!(error instanceof OperationCancelledError)) {
```
Same thought around inverting branches ` (!(error` readability vs `(error`

piotr-oles replied:
:+1:

johnnyreilly commented on src/reporter/FilesChange.ts:
```diff
@@ -0,0 +1,39 @@
+}
+
+function aggregateFilesChanges(changes: FilesChange[]): FilesChange {
+  const { nonExistedFiles, existedFiles } = aggregateFilesState(changes);
```
I'd be tempted to have different variable names here; `nonExistedFiles, existedFiles` are (if you'll excuse the opinion) not great English. ☺️

Suggest maybe `nonExistentFiles, existingFiles`? Or even `filesThatDontExist, filesThatExist`

piotr-oles replied:
haha, right 😄

johnnyreilly commented on src/reporter/FilesChange.ts:
```diff
@@ -0,0 +1,39 @@
+function aggregateFilesChanges(changes: FilesChange[]): FilesChange {
+  const { nonExistedFiles, existedFiles } = aggregateFilesState(changes);
+
+  return changes.reduce<FilesChange>(
```
I'm a man that loves a `.reduce`. Best thing ever ☺️

That said, I found this one hard to understand; I've read and re-read it and I still don't quite follow it. I wonder if this could be made simpler? Or if some comments could be added to clarify what's happening here?

piotr-oles replied:
I agree it's hard to read. I will try to make it more readable. Apart from that, I will comment on this fragment and add unit tests - the underlying logic is not trivial.

piotr-oles replied:
Could you check if it's more readable for you now? :)

johnnyreilly replied:
About a million times more readable 🥰

johnnyreilly commented on src/reporter/FilesState.ts:
```diff
@@ -0,0 +1,30 @@
+}
+
+function aggregateFilesState(changes: FilesChange[]): FilesState {
+  return changes.reduce<FilesState>(
```
Same thoughts as on `FilesChange.ts`

piotr-oles replied:
:+1:

johnnyreilly commented on src/reporter/index.ts:
```diff
@@ -0,0 +1,9 @@
+export * from './AggregatedReporter';
+
+export * from './Report';
+export * from '../error/OperationCancelledError';
```
Do you want to export `error/OperationCancelledError` from here given it lives in a sibling directory? Just wanted to check this was intentional?

piotr-oles replied:
no, it wasn't - thanks for catching this :)

johnnyreilly commented on src/rpc/rpc-ipc/RpcIpcMessagePort.ts:
```diff
@@ -0,0 +1,109 @@
+
+    closedError = new RpcIpcMessagePortClosedError(
+      `Process ${process.pid} exited with code ${code}.`,
+      code,
```
I'd be tempted to bin the `@typescript-eslint/no-use-before-define` rule - I've never been that convinced of the value it adds

piotr-oles replied:
:+1:

johnnyreilly commented on src/typescript-reporter/TypeScriptDiagnosticsOptions.ts:
```diff
@@ -0,0 +1,8 @@
+interface TypeScriptDiagnosticsOptions {
```
I think this is way cool!

johnnyreilly commented on src/typescript-reporter/assertTypeScriptSupport.ts:
```diff
@@ -0,0 +1,46 @@
+    );
+  }
+
+  if (semver.lt(typescriptVersion, '2.1.0')) {
```
Is TypeScript 2.1 is still the minimum version given the APIs being used?

piotr-oles replied:
I have to check that - it's on my TODO list as "Reevaluate supported versions" ;)

johnnyreilly commented on src/typescript-reporter/reporter/ControlledWatchSolutionBuilderHost.ts:
```diff
@@ -0,0 +1,107 @@
+      // return ts.sys.writeFile(fileName, data, writeByteOrderMark);
+    },
+    deleteFile(path: string): void {
+      // if (ts.sys.deleteFile) {
```
Is the commented code intentional?

piotr-oles replied:
Yes, it's temporary. I think I will add a configuration option to choose between "passive" and "active" mode for `build: true`. The default setting will be a "passive" mode in which the plugin doesn't write to the file system.  In this case, it will be a responsibility of the `ts-loader` or `babel-loader`

johnnyreilly replied:
This sounds interesting; could you tell me more about the use case for `build: true` and `build: false` please?

piotr-oles replied:
`build` flag is an equivalent to the `--build` flag in `tsc` command. When `build: true`, we use `SolutionBuilder` which consists of many `BuilderProgram`s (project references) in opposition to the `build: false` where we have only one `BuilderProgram`.  In the `build: true` mode TypeScript will try to build project references if they are not built or are not up to date. In the `build: false` mode TypeScript will throw an error if project references are not built and will not check if they are up to date with their source.
The `SolutionBuilder` API is available starting from TypeScript 3.6

johnnyreilly commented on src/typescript-reporter/reporter/ControlledWatchSolutionBuilderHost.ts:
```diff
@@ -0,0 +1,107 @@
+      }
+    },
+    setModifiedTime(path: string, date: Date): void {
+      // if (ts.sys.setModifiedTime) {
```
Intentional comments?

piotr-oles replied:
see comment above :)

johnnyreilly reviewed (commented):
Amazing work so far @piotr-oles!

cc @sheetalkamat and @andrewbranch - I thought you might be interested in the work Piotr has been putting into making `fork-ts-checker-webpack-plugin` project references and tsbuildinfo friendly ☺️

piotr-oles committed:
refactor: code-review requested changes

johnnyreilly commented:
Yeah this is looking really promising! 🤗

piotr-oles committed:
feat: upgrade min node version to 8

As Node 6 support ended in April 2019 and Node 8 ended in January 2020 we decided to drop support for Node 6

BREAKING CHANGE: Dropped support for Node 6

piotr-oles committed:
test: make existing unit tests work again

piotr-oles committed:
chore: change typescript version from 2.7.0 to 2.7.1

In the CI test matrix I had to change typescript version to 2.7.1
because 2.7.0 is not published on the npm

piotr-oles committed:
test: modify tests to be able to run them on typescript 2.7.1

piotr-oles committed:
chore: select patch version instead of major for typescript

piotr-oles committed:
test: add test and codeclimate configuration

piotr-oles committed:
fix: make project-references really working + e2e tests

piotr-oles committed:
feat: drop support for Node 8 and add Node 14 to the matrix

Node 8 is no longer maintained and we want to limit the number of
supported environments as it adds maintanance cost.

BREAKING CHANGE: 🧨 Dropped support for the Node 8

piotr-oles committed:
test: update tests for compatibility with windows

piotr-oles committed:
chore: update semantic-release configuration (#410)

Release new patch versions for "refactor" and "docs" commits.

andyrooger committed:
refactor(dependency): update babel-code-frame to @babel/code-frame (#409)

Update code-frame dependency from v6 to v7 so that we can drop the chalk v1 dependency.

Closes #408

piotr-oles committed:
test: stabilize e2e tests on linux and mac

piotr-oles committed:
test: stabilize e2e tests on windows

piotr-oles committed:
test: add tests for vue and pnp typescript extension

piotr-oles committed:
docs: update README.md to reflect new API

piotr-oles committed:
feat: add skipLibCheck compiler option as a default one

In order to improve developer experience, we set `skipDefaultLibCheck`
and `skipLibCheck` to true. It will improve type checking time.

piotr-oles committed:
test: add type definitions test

piotr-oles committed:
feat: add support for cosmiconfig

piotr-oles committed:
chore: add alpha branch to pre-release channels

piotr-oles committed:
feat: use developer-friendly compilerOptions

piotr-oles committed:
feat: improve support for the `incremental` mode

piotr-oles committed:
feat: support webpack-dev-server errors for async mode

Thanks @ypresto for the implementation of
https://github.com/ypresto/fork-ts-checker-async-overlay-webpack-plugin
- it was integrated with the plugin in this commit.

piotr-oles commented:
Hi @johnnyreilly , I implemented most of the points from the checklist, wrote end-to-end tests, and updated the documentation. Could you re-review this PR?

piotr-oles committed:
feat: add support for eslint fix in the new codebase

piotr-oles committed:
fix: resolve vue.js cache bug and add e2e tests

piotr-oles committed:
feat: improve message for vue.js compiler loading error

piotr-oles commented:
As this is an only alpha release and all tests passed, I will merge it :)

@johnnyreilly , @phryneas I know it's a big PR, so please review it when you have more time and if you have any comments, I will address them in the next alpha releases.

phryneas commented on src/index.ts:
```diff
@@ -1,726 +1,3 @@
-}
+import { ForkTsCheckerWebpackPlugin } from './ForkTsCheckerWebpackPlugin';
 
 export = ForkTsCheckerWebpackPlugin;
```
Wouldn't the new version be a great chance to finally make this a default or named export?

piotr-oles replied:
I would stick to the current export because it's done the same way in other webpack plugins (including webpack's official plugins) :)

phryneas replied:
That's a good point :)

phryneas commented on src/ForkTsCheckerWebpackPlugin.ts:
```diff
@@ -0,0 +1,68 @@
+  private readonly options: ForkTsCheckerWebpackPluginOptions;
+
+  constructor(options: ForkTsCheckerWebpackPluginOptions = {}) {
+    const explorerSync = cosmiconfigSync('fork-ts-checker');
```
Cool addition :)

phryneas commented on src/hooks/tapStopToDisconnectReporter.ts:
```diff
@@ -0,0 +1,52 @@
+    }
+  });
+
+  const hooks = getForkTsCheckerWebpackPluginHooks(compiler);
+
+  hooks.error.tap('ForkTsCheckerWebpackPlugin', (error) => {
+    configuration.logger.issues.error(String(error));
+
+    if (error instanceof RpcIpcMessagePortClosedError) {
+      if (error.signal === 'SIGINT') {
+        configuration.logger.issues.error(
+          chalk.red(
+            'Issues checking service interrupted - If running in a docker container, this may be caused ' +
+              "by the container running out of memory. If so, try increasing the container's memory limit " +
+              'or lowering the `memoryLimit` value in the ForkTsCheckerWebpackPlugin configuration.'
+          )
+        );
+      } else {
+        configuration.logger.issues.error(
+          chalk.red(
+            'Issues checking service aborted - probably out of memory. ' +
+              'Check the `memoryLimit` option in the ForkTsCheckerWebpackPlugin configuration.\n' +
+              "If increasing the memory doesn't solve the issue, it's most probably a bug in the TypeScript or EsLint."
+          )
+        );
+      }
+    }
```
This might make more logical sense as an own function - or even file. It doesn't really tap into `stop` and also it doesn't disconnect the `reporter`.

piotr-oles replied:
:+1:

piotr-oles replied:
Addressed in the #417

phryneas commented on src/hooks/getChangedFiles.ts:
```diff
@@ -0,0 +1,31 @@
+function getChangedFiles(compiler: webpack.Compiler): string[] {
+  let changedFiles: string[] = [];
+
+  if ((compiler as any).modifiedFiles) {
```
Generally, there is a lot of `compiler as any` going on and that makes some stuff hard to read. Maybe it would make sense to define some interface extending `webpack.Compiler` and cast that at a central location once or twice?

Something like 
```
type Webpack4Or5 = webpack.Compiler & ({
// webpack 5 stuff
} | {
// webpack 4 stuff
})
```

The compiler should be able to use the relevant properties as discriminator from there.

piotr-oles replied:
I will try that :)

phryneas commented on src/typescript-reporter/extension/pnp/TypeScriptPnpExtension.ts:
```diff
@@ -0,0 +1,52 @@
+import * as ts from 'typescript';
+import { TypeScriptExtension } from '../TypeScriptExtension';
+
+function createTypeScriptPnpExtension(): TypeScriptExtension {
```
Is this still the right way to do this? Afaik, the `pnpify`-patched `typescript` should not need any additional configuration in `fork-ts-checker-webpack-plugin`, so providing the `TypeScriptPnpExtension` could be more confusing than helpful to end-users.
I might be wrong though. @arcanis to the rescue?

arcanis replied:
I think this part of the code isn't required anymore since we now add the resolution pipeline directly within TypeScript, indeed!

piotr-oles replied:
Good, then I will remove this code :)

piotr-oles replied:
Addressed in the #418

phryneas commented on src/typescript-reporter/reporter/TypeScriptReporter.ts:
```diff
@@ -0,0 +1,171 @@
+
+  const diagnosticsPerProject = new Map<string, ts.Diagnostic[]>();
+
+  if (configuration.extensions.vue.enabled) {
+    extensions.push(createTypeScriptVueExtension(configuration.extensions.vue));
+  }
+  if (configuration.extensions.pnp.enabled) {
+    extensions.push(createTypeScriptPnpExtension());
+  }
```
Could this be a good moment to move those extensions out to separate packages and make them injectable via configuration?

piotr-oles replied:
I would keep them in the same package for now. In my opinion, these extensions are a workaround and I hope that the TypeScript core team will agree to add API for compiler plugins. Currently we need to implement these extensions in the plugin, loader, and language service. I've created https://github.com/microsoft/TypeScript/issues/38736 proposal based on the experience with the plugin - maybe it will move forward. If they will not agree, then we can think about a unified way to write TypeScript plugins that could work both in the plugin and ts-loader :)

phryneas replied:
I'm rooting for it. Let's see how it goes :)

phryneas reviewed (commented):
This is huuuge :smile:
Congrats to that big hunk of work you pulled off there - very impressive!

Is's not really possible to review this as a diff, so I just took a deep dive at the source code.

I really like the new architecture, it's a lot cleaner - great job on that :+1:  
After reading it, I have to say: I understand what's going on, but if there are any bugs, I probably missed those because I'm not too accustomed to the code base and also haven't had a lot of time to spend with webpack in the recent past. So my review comments are primarily of a more generic nature.

I'll definitely start dogfooding this as soon as I get back into our react project (doing a lot of infra at the moment) and will report back if I find any anomalies there.

johnnyreilly commented:
Just wanted to share these timings that @timer was kind enough to do here: https://github.com/vercel/next.js/pull/13529#issuecomment-635776458

> Base build: 7s (no TypeScript features enabled)
>
> `fork-ts-checker-webpack-plugin@3.1.1`: 90s, computer sounds like an airplane
> `fork-ts-checker-webpack-plugin@4.1.6`: 84s, computer did not sound like an airplane
> `fork-ts-checker-webpack-plugin@5.0.0-alpha.14`: 90s, regressed
> `npx tsc -p tsconfig.json --noEmit`: 12s (time: 18.57s user 0.97s system 169% cpu 11.525 total)

weyert commented:
John, do you think this is due the rewrite to TypeScript or has more changed during the rewrite? I can imagine the TypeScript transpilling does some more runtime checks

piotr-oles commented:
:tada: This PR is included in version 5.0.0-alpha.17 :tada:

The release is available on:
- [npm package (@alpha dist-tag)](https://www.npmjs.com/package/fork-ts-checker-webpack-plugin/v/5.0.0-alpha.17)
- [GitHub release](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/releases/tag/v5.0.0-alpha.17)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

piotr-oles commented:
:tada: This PR is included in version 5.0.0-beta.2 :tada:

The release is available on:
- [npm package (@beta dist-tag)](https://www.npmjs.com/package/fork-ts-checker-webpack-plugin/v/5.0.0-beta.2)
- [GitHub release](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/releases/tag/v5.0.0-beta.2)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
